controllers/securitygroup/securitygroup_controller.go

```diff
@@ -69,71 +71,200 @@ const (
+		Scheme:                          mgr.GetScheme(),
+		Recorder:                        mgr.GetEventRecorderFor("securityGroup-controller"),
+		NewEC2ClientFactory:             ec2.NewEC2Client,
+		NewAutoScalingClientFactory:     autoscaling.NewAutoScalingClient,
+		NewOceanCloudProviderAWSFactory: spot.NewOceanCloudProviderAWS,
+	}
+}
+
+func awsConfigForCredential(ctx context.Context, region string, accessKey string, secretAccessKey string) (aws.Config, error) {
+	return config.LoadDefaultConfig(ctx,
+		config.WithRegion(region),
+		config.WithCredentialsProvider(credentials.StaticCredentialsProvider{
+			Value: aws.Credentials{
+				AccessKeyID:     accessKey,
+				SecretAccessKey: secretAccessKey,
+			},
+		}),
+	)
+}
 
-	sg := &securitygroupv1alpha1.SecurityGroup{}
-	if err := r.Get(ctx, req.NamespacedName, sg); err != nil {
+func (c *SecurityGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, rerr error) {
+	r := &SecurityGroupReconciliation{
+		SecurityGroupReconciler: *c,
+		start:                   time.Now(),
+		log:                     ctrl.LoggerFrom(ctx),
+	}
+
+	r.sg = &securitygroupv1alpha1.SecurityGroup{}
+	if err := r.Get(ctx, req.NamespacedName, r.sg); err != nil {
 		return resultError, client.IgnoreNotFound(err)
 	}
 
-	if sg.Spec.InfrastructureRef == nil {
-		return resultError, fmt.Errorf("infrastructureRef isn't defined")
+	if r.sg.Spec.InfrastructureRef == nil {
+		return resultDefault, fmt.Errorf("infrastructureRef not supported")
+	}
+
+	switch r.sg.Spec.InfrastructureRef.Kind {
+	case "KopsMachinePool":
+		r.kmp = &kinfrastructurev1alpha1.KopsMachinePool{}
+		key := client.ObjectKey{
+			Name:      r.sg.Spec.InfrastructureRef.Name,
+			Namespace: r.sg.Spec.InfrastructureRef.Namespace,
+		}
+		if err := r.Client.Get(ctx, key, r.kmp); err != nil {
+			return resultError, err
+		}
+
+		r.kcp = &kcontrolplanev1alpha1.KopsControlPlane{}
+		key = client.ObjectKey{
+			Name:      r.kmp.Spec.ClusterName,
+			Namespace: r.kmp.ObjectMeta.Namespace,
+		}
+		if err := r.Client.Get(ctx, key, r.kcp); err != nil {
+			return resultError, err
+		}
+	case "KopsControlPlane":
+		r.kcp = &kcontrolplanev1alpha1.KopsControlPlane{}
+		key := client.ObjectKey{
+			Name:      r.sg.Spec.InfrastructureRef.Name,
+			Namespace: r.sg.Spec.InfrastructureRef.Namespace,
+		}
+		if err := r.Client.Get(ctx, key, r.kcp); err != nil {
+			return resultError, err
+		}
+	default:
+		return resultError, fmt.Errorf("infrastructureRef not supported")
+	}
+
+	subnet, err := kops.GetSubnetFromKopsControlPlane(r.kcp)
+	if err != nil {
+		r.log.Error(err, fmt.Sprintf("failed to get subnet from kcp"))
+		return resultError, err
+	}
+
+	region, err := kops.GetRegionFromKopsSubnet(*subnet)
+	if err != nil {
+		r.log.Error(err, fmt.Sprintf("failed to get region from kcp"))
+		return resultError, err
+	}
+
+	var secretName, namespace string
+	if r.kcp.Spec.IdentityRef != nil {
+		secretName = r.kcp.Spec.IdentityRef.Name
+		// gambiarra??
```

Comment:
is this bad? Maybe I should just error if it is not specified?

Reply:
I agree, is better to return error if this isn't specified